File: ContextMcp/src/review/rules/securityRules.ts
import ts from "typescript"
import type { Finding } from "../finding"
import type { TsRule } from "../tsRule"
import { lineOf, walk } from "../tsRule"

const SECRET_PATTERNS: Array<{ regex: RegExp; label: string }> = [
  { regex: /AKIA[0-9A-Z]{16}/, label: "AWS access key" },
  { regex: /-----BEGIN (?:RSA |EC |DSA |OPENSSH |PGP )?PRIVATE KEY-----/, label: "private key (PEM)" },
  { regex: /eyJ[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}/, label: "JWT token" },
  { regex: /(?:password|passwd|pwd|api[_-]?key|secret|token)\s*[:=]\s*['"][^'"\s]{6,}['"]/i, label: "credential literal" },
  { regex: /(?:mongodb|postgres|postgresql|mysql|mssql|redis):\/\/[^\s'"<>]+:[^\s'"<>]+@/i, label: "DB connection string with credentials" },
]

const hardcodedSecret: TsRule = {
  name: "hardcoded-secret",
  category: "Security",
  check(sf, fileName) {
    const out: Finding[] = []
    walk(sf, (node) => {
      if (
        ts.isStringLiteral(node) ||
        ts.isNoSubstitutionTemplateLiteral(node) ||
        ts.isTemplateHead(node) ||
        ts.isTemplateMiddle(node) ||
        ts.isTemplateTail(node)
      ) {
        const text = node.text
        for (const { regex, label } of SECRET_PATTERNS) {
          if (regex.test(text)) {
            out.push({
              severity: "Critical",
              category: "Security",
              file: fileName,
              line: lineOf(sf, node.getStart(sf)),
              rule: "hardcoded-secret",
              message: `Olası hardcoded ${label} string literal içinde.`,
              impact:
                "Repo'ya commit edilen sırlar git history'sinde kalır; ihlali geri alınamaz, " +
                "credential rotasyonu ve incident response gerektirir.",
              recommendation:
                "Sırrı environment variable veya secret manager'a taşı (örn: process.env.X, " +
                "AWS Secrets Manager, Vault). Repo'dan kaldırıp anahtarı rotate et.",
            })
            break
          }
        }
      }
    })
    return out
  },
}

const SQL_KEYWORD = /\b(SELECT|INSERT|UPDATE|DELETE|FROM|WHERE|JOIN|UNION|DROP|ALTER|TRUNCATE)\b/i

const sqlInjection: TsRule = {
  name: "sql-injection",
  category: "Security",
  check(sf, fileName) {
    const out: Finding[] = []
    walk(sf, (node) => {
      if (ts.isTemplateExpression(node)) {
        const headText = node.head.text
        const allText =
          headText + node.templateSpans.map((s) => s.literal.text).join(" ")
        if (SQL_KEYWORD.test(allText) && node.templateSpans.length > 0) {
          out.push({
            severity: "Critical",
            category: "Security",
            file: fileName,
            line: lineOf(sf, node.getStart(sf)),
            rule: "sql-injection",
            message: "Template literal SQL anahtar kelimesi içeriyor ve interpolasyon yapıyor.",
            impact:
              "Kullanıcı girdisinin SQL'e birleştirilmesi injection açar; " +
              "yetkisiz veri okuma, silme veya tüm tablo dökümünü mümkün kılar.",
            recommendation:
              "Parametrik sorgu kullan: pg/mysql2 `?`/`$1` placeholder'ları, " +
              "Prisma/TypeORM parameterized queries veya stored procedure. " +
              "Sorgu içeriği kullanıcıdan gelmemeli.",
          })
        }
      }

      if (ts.isBinaryExpression(node) && node.operatorToken.kind === ts.SyntaxKind.PlusToken) {
        const left = node.left.getText(sf)
        const right = node.right.getText(sf)
        const combined = `${left} ${right}`
        if (SQL_KEYWORD.test(combined) && (left.includes('"') || left.includes("'") || right.includes('"') || right.includes("'"))) {
          if (!ts.isStringLiteralLike(node.left) || !ts.isStringLiteralLike(node.right)) {
            out.push({
              severity: "Critical",
              category: "Security",
              file: fileName,
              line: lineOf(sf, node.getStart(sf)),
              rule: "sql-injection",
              message: "SQL sorgusu string concatenation (+) ile değişken birleştiriyor.",
              impact:
                "String concat ile inşa edilen SQL injection açar; yetkisiz erişim ve veri kaybı riski.",
              recommendation:
                "Parametrik sorgu (placeholder) kullan veya ORM'in parametre bağlama API'sini kullan.",
            })
          }
        }
      }
    })
    return out
  },
}

const xssSink: TsRule = {
  name: "xss-sink",
  category: "Security",
  check(sf, fileName) {
    const out: Finding[] = []
    walk(sf, (node) => {
      if (
        ts.isBinaryExpression(node) &&
        node.operatorToken.kind === ts.SyntaxKind.EqualsToken &&
        ts.isPropertyAccessExpression(node.left) &&
        node.left.name.text === "innerHTML" &&
        !ts.isStringLiteral(node.right)
      ) {
        out.push({
          severity: "High",
          category: "Security",
          file: fileName,
          line: lineOf(sf, node.getStart(sf)),
          rule: "xss-sink",
          message: "innerHTML'e statik olmayan değer atanıyor.",
          impact:
            "Kullanıcı kontrolündeki içerik innerHTML'e yazılırsa stored/reflected XSS açar; " +
            "session hijack, credential theft, defacement mümkün olur.",
          recommendation:
            "textContent kullan veya DOMPurify gibi sanitizer'dan geçir. " +
            "React'te dangerouslySetInnerHTML yerine JSX kullan.",
        })
      }

      if (
        ts.isJsxAttribute(node) &&
        node.name.getText(sf) === "dangerouslySetInnerHTML"
      ) {
        out.push({
          severity: "High",
          category: "Security",
          file: fileName,
          line: lineOf(sf, node.getStart(sf)),
          rule: "xss-sink",
          message: "JSX'te dangerouslySetInnerHTML kullanılıyor.",
          impact:
            "React'in XSS koruması bypass edilir; içerik sanitize edilmezse XSS açar.",
          recommendation:
            "Mümkünse JSX kullan. Zorunluysa DOMPurify ile sanitize et ve yorumda gerekçesini açıkla.",
        })
      }
    })
    return out
  },
}

const weakCrypto: TsRule = {
  name: "weak-crypto",
  category: "Security",
  check(sf, fileName) {
    const out: Finding[] = []
    walk(sf, (node) => {
      if (
        ts.isCallExpression(node) &&
        ts.isPropertyAccessExpression(node.expression) &&
        node.expression.name.text === "createHash" &&
        node.arguments.length > 0 &&
        ts.isStringLiteralLike(node.arguments[0])
      ) {
        const algo = node.arguments[0].text.toLowerCase()
        if (algo === "md5" || algo === "sha1") {
          out.push({
            severity: "High",
            category: "Security",
            file: fileName,
            line: lineOf(sf, node.getStart(sf)),
            rule: "weak-crypto",
            message: `Zayıf hash algoritması: ${algo.toUpperCase()}.`,
            impact:
              "MD5/SHA1 collision saldırılarına açık; password/integrity için kullanılırsa " +
              "preimage bulma ve manipülasyon mümkündür.",
            recommendation:
              "SHA-256 veya SHA-512 kullan (createHash('sha256')). Şifre hashleme için " +
              "bcrypt/scrypt/argon2id tercih et.",
          })
        }
      }
    })
    return out
  },
}

const unsafeEval: TsRule = {
  name: "unsafe-eval",
  category: "Security",
  check(sf, fileName) {
    const out: Finding[] = []
    walk(sf, (node) => {
      if (ts.isCallExpression(node) && ts.isIdentifier(node.expression) && node.expression.text === "eval") {
        out.push({
          severity: "Critical",
          category: "Security",
          file: fileName,
          line: lineOf(sf, node.getStart(sf)),
          rule: "unsafe-eval",
          message: "eval() kullanılıyor.",
          impact:
            "eval() kullanıcı girdisini çalıştırırsa keyfi kod yürütme (RCE) açar; " +
            "ayrıca optimizer'ları engeller ve scope analizini bozar.",
          recommendation:
            "JSON.parse, Function constructor (sınırlı senaryolarda), parser kütüphanesi " +
            "veya tasarımı yeniden değerlendir — kullanıcı girdisi eval edilmemeli.",
        })
      }

      if (ts.isNewExpression(node) && ts.isIdentifier(node.expression) && node.expression.text === "Function") {
        out.push({
          severity: "Critical",
          category: "Security",
          file: fileName,
          line: lineOf(sf, node.getStart(sf)),
          rule: "unsafe-eval",
          message: "new Function() ile dinamik kod yürütülüyor.",
          impact:
            "eval ile aynı RCE riski. Kullanıcı girdisi Function constructor'a geçerse " +
            "keyfi kod yürütülür.",
          recommendation:
            "Dinamik mantığı static dispatch (switch/map) ile değiştir veya güvenli sandbox kullan.",
        })
      }
    })
    return out
  },
}

export const SECURITY_RULES: TsRule[] = [
  hardcodedSecret,
  sqlInjection,
  xssSink,
  weakCrypto,
  unsafeEval,
]

File: ContextMcp.Roslyn/Review/Rules/SecurityRules.cs
using System.Text.RegularExpressions;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;

namespace ContextMcp.Roslyn.Review.Rules;

internal sealed class HardcodedSecretRule : IRule
{
    public string Name => "hardcoded-secret";
    public string Category => "Security";

    private static readonly (Regex regex, string label)[] Patterns = new (Regex, string)[]
    {
        (new Regex(@"AKIA[0-9A-Z]{16}", RegexOptions.Compiled), "AWS access key"),
        (new Regex(@"-----BEGIN (?:RSA |EC |DSA |OPENSSH |PGP )?PRIVATE KEY-----", RegexOptions.Compiled), "private key (PEM)"),
        (new Regex(@"eyJ[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}", RegexOptions.Compiled), "JWT token"),
        (new Regex(@"(?:password|passwd|pwd|api[_-]?key|secret|token)\s*[:=]\s*[""'][^""'\s]{6,}[""']", RegexOptions.Compiled | RegexOptions.IgnoreCase), "credential literal"),
        (new Regex(@"(?:mongodb|postgres|postgresql|mysql|mssql|redis)://[^\s'""<>]+:[^\s'""<>]+@", RegexOptions.Compiled | RegexOptions.IgnoreCase), "DB connection string with credentials"),
    };

    public IEnumerable<Finding> Check(SyntaxTree tree, SemanticModel model, string filePath)
    {
        var root = tree.GetRoot();
        foreach (var literal in root.DescendantNodes().OfType<LiteralExpressionSyntax>())
        {
            if (!literal.IsKind(SyntaxKind.StringLiteralExpression)) continue;
            var text = literal.Token.ValueText;
            foreach (var (regex, label) in Patterns)
            {
                if (regex.IsMatch(text))
                {
                    yield return new Finding(
                        "Critical", "Security", filePath, RuleHelpers.LineOf(literal),
                        Name,
                        $"Olası hardcoded {label} string literal içinde.",
                        "Repo'ya commit edilen sırlar git history'sinde kalır; geri alınamaz, " +
                        "credential rotasyonu ve incident response gerektirir.",
                        "Sırrı environment variable veya secret manager'a (User Secrets, Azure Key Vault, " +
                        "AWS Secrets Manager) taşı; repo'dan kaldırıp anahtarı rotate et.");
                    break;
                }
            }
        }
    }
}

internal sealed class SqlInjectionRule : IRule
{
    public string Name => "sql-injection";
    public string Category => "Security";

    private static readonly Regex SqlKeyword = new(
        @"\b(SELECT|INSERT|UPDATE|DELETE|FROM|WHERE|JOIN|UNION|DROP|ALTER|TRUNCATE)\b",
        RegexOptions.Compiled | RegexOptions.IgnoreCase);

    public IEnumerable<Finding> Check(SyntaxTree tree, SemanticModel model, string filePath)
    {
        var root = tree.GetRoot();

        // 1) Interpolated string SQL keyword içeriyor → injection riski
        foreach (var interp in root.DescendantNodes().OfType<InterpolatedStringExpressionSyntax>())
        {
            var rawText = interp.ToString();
            var hasInterpolation = interp.Contents.OfType<InterpolationSyntax>().Any();
            if (hasInterpolation && SqlKeyword.IsMatch(rawText))
            {
                yield return new Finding(
                    "Critical", "Security", filePath, RuleHelpers.LineOf(interp),
                    Name,
                    "Interpolated string SQL anahtar kelimesi içeriyor ve değişken interpolasyon yapıyor.",
                    "Kullanıcı girdisinin SQL'e birleştirilmesi injection açar; yetkisiz veri okuma, " +
                    "silme veya tüm tablo dökümünü mümkün kılar.",
                    "Parametre kullan: SqlCommand.Parameters.AddWithValue(...), " +
                    "Dapper named parameters, EF Core .FromSqlInterpolated yerine LINQ.");
            }
        }

        // 2) String concat (+) ile SQL inşası
        foreach (var bin in root.DescendantNodes().OfType<BinaryExpressionSyntax>())
        {
            if (!bin.IsKind(SyntaxKind.AddExpression)) continue;
            var text = bin.ToString();
            if (!SqlKeyword.IsMatch(text)) continue;

            // En az bir tarafı non-literal mi?
            bool leftLiteral = bin.Left is LiteralExpressionSyntax;
            bool rightLiteral = bin.Right is LiteralExpressionSyntax;
            if (leftLiteral && rightLiteral) continue;

            yield return new Finding(
                "Critical", "Security", filePath, RuleHelpers.LineOf(bin),
                Name,
                "SQL sorgusu string concatenation (+) ile değişken birleştiriyor.",
                "String concat ile inşa edilen SQL injection açar; yetkisiz erişim ve veri kaybı riski.",
                "Parametrik sorgu kullan veya ORM'in parametre bağlama API'sini kullan " +
                "(Dapper, EF Core, SqlParameter).");
        }
    }
}

internal sealed class WeakCryptoRule : IRule
{
    public string Name => "weak-crypto";
    public string Category => "Security";

    private static readonly HashSet<string> WeakAlgos = new(StringComparer.OrdinalIgnoreCase)
    {
        "MD5", "SHA1", "MD5CryptoServiceProvider", "SHA1CryptoServiceProvider", "SHA1Managed", "MD5Managed",
    };

    public IEnumerable<Finding> Check(SyntaxTree tree, SemanticModel model, string filePath)
    {
        var root = tree.GetRoot();

        // 1) MD5.Create() / SHA1.Create() statik çağrı
        foreach (var invocation in root.DescendantNodes().OfType<InvocationExpressionSyntax>())
        {
            if (invocation.Expression is MemberAccessExpressionSyntax ma &&
                ma.Name.Identifier.Text == "Create" &&
                ma.Expression is IdentifierNameSyntax id &&
                WeakAlgos.Contains(id.Identifier.Text))
            {
                yield return new Finding(
                    "High", "Security", filePath, RuleHelpers.LineOf(invocation),
                    Name,
                    $"Zayıf hash algoritması: {id.Identifier.Text}.Create().",
                    "MD5/SHA1 collision saldırılarına açık; password veya integrity için kullanılırsa " +
                    "preimage bulma ve manipülasyon mümkündür.",
                    "SHA256.Create() veya daha güçlüsünü kullan. Şifre hashleme için " +
                    "Microsoft.AspNetCore.Identity.PasswordHasher (PBKDF2) veya BCrypt/Argon2 tercih et.");
            }
        }

        // 2) new MD5CryptoServiceProvider() benzeri
        foreach (var obj in root.DescendantNodes().OfType<ObjectCreationExpressionSyntax>())
        {
            var typeName = obj.Type.ToString();
            if (WeakAlgos.Contains(typeName))
            {
                yield return new Finding(
                    "High", "Security", filePath, RuleHelpers.LineOf(obj),
                    Name,
                    $"Zayıf hash algoritması: new {typeName}().",
                    "MD5/SHA1 cryptographically broken; collision saldırıları pratik.",
                    "SHA256/SHA512 kullan veya şifre için PBKDF2/Argon2.");
            }
        }
    }
}

internal sealed class MissingAuthorizeRule : IRule
{
    public string Name => "missing-authorize";
    public string Category => "Security";

    private static readonly string[] HttpVerbAttrs = { "HttpGet", "HttpPost", "HttpPut", "HttpDelete", "HttpPatch" };

    public IEnumerable<Finding> Check(SyntaxTree tree, SemanticModel model, string filePath)
    {
        var root = tree.GetRoot();

        foreach (var cls in root.DescendantNodes().OfType<ClassDeclarationSyntax>())
        {
            // Controller mı?
            var baseTypes = cls.BaseList?.Types.Select(t => t.Type.ToString()).ToList() ?? new();
            bool isController =
                baseTypes.Contains("ControllerBase") ||
                baseTypes.Contains("Controller") ||
                cls.Identifier.Text.EndsWith("Controller");
            if (!isController) continue;

            bool classHasAuth = HasAuthAttribute(cls.AttributeLists) || HasAllowAnonymous(cls.AttributeLists);

            foreach (var method in cls.Members.OfType<MethodDeclarationSyntax>())
            {
                if (!IsPublicMethod(method)) continue;

                bool hasHttpVerb = method.AttributeLists
                    .SelectMany(al => al.Attributes)
                    .Any(a => HttpVerbAttrs.Contains(a.Name.ToString()));
                if (!hasHttpVerb) continue;

                bool methodHasAuth = HasAuthAttribute(method.AttributeLists) || HasAllowAnonymous(method.AttributeLists);

                if (!classHasAuth && !methodHasAuth)
                {
                    yield return new Finding(
                        "High", "Security", filePath, RuleHelpers.LineOf(method),
                        Name,
                        $"Action '{method.Identifier.Text}' [Authorize] / [AllowAnonymous] olmadan public endpoint.",
                        "Erişim kontrolü olmayan endpoint kimliği doğrulanmamış tüm trafiğe açıktır; " +
                        "veri sızıntısı, yetkisiz iş emri, IDOR riski.",
                        "Controller veya action'a [Authorize(...)] ekle; bilinçli olarak public ise " +
                        "[AllowAnonymous] ile niyetini ifade et.");
                }
            }
        }
    }

    private static bool HasAuthAttribute(SyntaxList<AttributeListSyntax> lists) =>
        lists.SelectMany(al => al.Attributes).Any(a =>
        {
            var n = a.Name.ToString();
            return n == "Authorize" || n.EndsWith(".Authorize");
        });

    private static bool HasAllowAnonymous(SyntaxList<AttributeListSyntax> lists) =>
        lists.SelectMany(al => al.Attributes).Any(a =>
        {
            var n = a.Name.ToString();
            return n == "AllowAnonymous" || n.EndsWith(".AllowAnonymous");
        });

    private static bool IsPublicMethod(MethodDeclarationSyntax method) =>
        method.Modifiers.Any(m => m.Text == "public");
}

internal static class SecurityRules
{
    public static readonly IRule[] All = new IRule[]
    {
        new HardcodedSecretRule(),
        new SqlInjectionRule(),
        new WeakCryptoRule(),
        new MissingAuthorizeRule(),
    };
}
